components/function-controller/pkg/apis/serverless/v1alpha2/function_types.go

```diff
@@ -118,6 +120,67 @@ type ResourceConfiguration struct {
 	Function *ResourceRequirements `json:"function,omitempty"`
 }
 
+type PodSpecTemplate struct {
+	// Compute Resources required by container created by function controller.
+	// +optional
+	Resources *ResourceRequirements `json:"resources,omitempty"`
+
+	// List of environment variables to set in the container.
+	// Cannot be updated.
+	// +optional
+	// +patchMergeKey=name
+	// +patchStrategy=merge
+	Env []v1.EnvVar `json:"env,omitempty"`
+
+	// Pod volumes to mount into the container's filesystem.
+	// +optional
+	// +patchMergeKey=mountPath
+	// +patchStrategy=merge
+	VolumeMounts []v1.VolumeMount `json:"volumeMounts,omitempty"`
+}
+
+type MetadataTemplate struct {
+	// Additional labels
+	// +optional
+	Labels map[string]string `json:"labels,omitempty"`
+
+	// Additional annotations
+	// +optional
+	Annotations map[string]string `json:"annotations,omitempty"`
+}
+
+type PodTemplate struct {
+	// Standard object's metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#metadata
+	// +optional
+	Metadata *MetadataTemplate `json:"metadata,omitempty"`
+
+	// Additional specification for pod created by function-controller
+	//+optional
+	Spec *PodSpecTemplate `json:"spec,omitempty"`
+}
+
+type Templates struct {
+	// Standard object's metadata.
+	// This metadata is applied on functionPod and buildPod.
+	// Can be overwritten by specific pod metadata
+	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#metadata
+	// +optional
+	Metadata *MetadataTemplate `json:"metadata,omitempty"`
```

Comment:
what is the purpose of metadata on this level ?

Reply:
To not repeat annotations and labels for build's pod and function's pods

Reply:
but we should repeat them, right? the main goal of creating the `templates`  field is to support config separation of build and function pods. Or maybe I misunderstood something...

Reply:
ok, so for shared ones?

Reply:
Yes, it's used for shared ones to not create very complicated structure and duplicate data. 
If user want to set specific labels for function's pod or build's pod, the user is able to customize it.

Reply:
I will remove this field.

Reply:
@dbadura why? It made sense.

Reply:
@moelsayed 
IMO we have complicated logic for calculating some fields - merging, defaulting, ... It complicate our code and do it harder for maintain. 

This field complicate our code because we should merge it with `metadata` inside templates: `buildJobTemplate` and `functionPodTemplate`. Remove it in the future needs change API version. Maybe users don't need this functionality now.

When we don't add it now, code is simpler. We can add this in every moment.